jobs/integration/test_opa.py

```diff
@@ -0,0 +1,439 @@
+import asyncio
+import json
+import random
+from pathlib import Path
+
+from juju.utils import block_until_with_coroutine
+import pytest
+
+from .logger import log
+from .utils import (
+    JujuRunError,
+    kubectl,
+    wait_for_application_status,
+    kubectl_apply,
+    kubectl_delete,
+)
+
+templates = Path(__file__).parent / "templates"
+
+
+class OPATestBase:
+    @pytest.fixture(autouse=True)
+    def setup(self, model, k8s_model, tools):
+        self.model = model
+        self.k8s_model = k8s_model
+        self.tools = tools
+        control_plane_app = self.model.applications["kubernetes-control-plane"]
+        self.control_plane_unit = control_plane_app.units[0]
+
+    async def validate_create_ns_with_label(self):
+        # Create the namespace
+        log("Creating namespace")
+        await kubectl_apply(
+            templates / "validate-gatekeeper-policy.yaml",
+            self.control_plane_unit,
+            self.tools,
+            self.model,
+        )
+
+        # Wait for namespace to go active
+        cmd = "get ns test-ns -o jsonpath='{.status.phase}'"
+        log("Waiting for namespace to become active")
+        while result := await kubectl(self.model, cmd):
+            if result.output == "Active":
+                break
+            await asyncio.sleep(1)
+
+        # Delete the namespace
+        cmd = "delete ns test-ns"
+        await kubectl_delete(
+            templates / "validate-gatekeeper-policy.yaml",
+            self.control_plane_unit,
+            self.tools,
+            self.model,
+        )
+
+    async def validate_create_ns(self, name=None):
+        if not name:
+            name = f"test-ns-{random.randint(1, 99999)}"
+
+        cmd = f"create ns {name}"
+        log("Creating namespace without any labels")
+        await kubectl(self.model, cmd)
+
+        # Wait for namespace to go active
+        cmd = f"get ns {name} -o jsonpath='{{.status.phase}}'"
+        log("Waiting for namespace to become active")
+        while result := await kubectl(self.model, cmd):
+            if result.output == "Active":
+                break
+            await asyncio.sleep(1)
+
+        # Delete the namespace
+        cmd = f"delete ns {name}"
+        await kubectl(self.model, cmd)
+
+    async def validate_create_ns_fail(self, name=None):
+        if not name:
+            name = f"test-ns-{random.randint(1, 99999)}"
+
+        cmd = f"create ns {name}"
+        # Creating the namespace should raise
+        log("Creating namespace without any labels")
+        try:
```

Comment:
[pytest](https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest-raises) tells me that testing for expected exceptions should look like this:

```python
with pytest.raises(JujuRunError) as ie:
    await kubectl(self.model, cmd)
```

now you've got an interesting wrinkle in here that says to retry for transient errors.  

So how about this, it's still a bit repetitive but it makes use of the `match` method in pytest.raises

```python
with pytest.raises(JujuRunError, match="Forbidden|InternalError") as exc_info:
    await kubectl(self.model, cmd)
    
if "InternalError" in str(exc_info.value):
    await asyncio.sleep(10)
    with pytest.raises(JujuRunError, match="Forbidden") as exc_info:
        await kubectl(self.model, cmd)
        
assert exc_info.output.startswith("Error from server (Forbidden) ...")
```